test/Microsoft.AspNet.Mvc.ModelBinding.Test/BodyBindingAndValidationIntegrationTest.cs

```diff
@@ -0,0 +1,186 @@
+
+        private class Address
+        {
+            public string Street { get; set; }
+        }
+
+        [Fact]
+        public async Task BodyBoundOnProperty_RequiredOnProperty_AddsModelStateError()
+        {
+            // Arrange
+            var argumentBinder = ModelBindingTestHelper.GetArgumentBinder();
+            var parameter = new ParameterDescriptor()
+            {
+                Name = "Parameter1",
+                BindingInfo = new BindingInfo()
+                {
+                    BinderModelName = "CustomParameter",
+                },
+                ParameterType = typeof(Person)
+            };
+
+            var operationContext = ModelBindingTestHelper.GetOperationBindingContext();
+            var httpContext = operationContext.HttpContext;
+            ConfigureHttpRequest(httpContext.Request, string.Empty);
+            var modelState = new ModelStateDictionary();
+
+            // Act
+            var model = await argumentBinder.BindModelAsync(parameter, modelState, operationContext);
+
+            // Assert
+            Assert.Equal("The Address field is required.", modelState[""].Errors.Single().ErrorMessage);
+        }
+
+        private class Person4
+        {
+            [FromBody]
+            [Required]
+            public int Address { get; set; }
+        }
+
+        [Fact]
+        public async Task BodyBoundOnProperty_RequiredOnValueTypeProperty_AddsModelStateError()
+        {
+            // Arrange
+            var argumentBinder = ModelBindingTestHelper.GetArgumentBinder();
+            var parameter = new ParameterDescriptor()
+            {
+                Name = "Parameter1",
+                BindingInfo = new BindingInfo()
+                {
+                    BinderModelName = "CustomParameter",
+                },
+                ParameterType = typeof(Person4)
+            };
+
+            var operationContext = ModelBindingTestHelper.GetOperationBindingContext();
+            var httpContext = operationContext.HttpContext;
+            ConfigureHttpRequest(httpContext.Request, string.Empty);
+            var modelState = new ModelStateDictionary();
+
+            // Act
+            var model = await argumentBinder.BindModelAsync(parameter, modelState, operationContext);
+
+            // Assert
+            Assert.StartsWith("No JSON content found and type 'System.Int32' is not nullable.",
+                modelState[""].Errors.Single().Exception.Message);
+        }
+
+        private class Person2
+        {
+            [FromBody]
+            public Address2 Address { get; set; }
+        }
+
+        private class Address2
+        {
+            [Required]
+            public string Street { get; set; }
+
+            public int Zip { get; set; }
+        }
+
+        [Theory]
+        [InlineData("{ \"Zip\" : 123 }")]
+        [InlineData("{}")]
+        public async Task BodyBoundOnTopLevelProperty_RequiredOnSubProperty_AddsModelStateError(string inputText)
```

Comment:
because the attribute is `[FromBody]` and "body bound" is a random term you're inventing